File: api/queries/comments.js
let { db } = require('../database');
import format from 'pg-format';


function createComment(req, res, next) {

  const now = new Date();

  body = {
    content: req.body.content ? req.body.content : null,
    post_id: req.body.post_id,
    author_id: req.user.id,
    created: now
  };

  if (!body.content) {

    res.status(400)
      .json({
        status: 'error',
        message: 'content can not be null.'
      });

  } else if (!body.post_id) {

    res.status(400)
      .json({
        status: 'error',
        message: 'post_id can not be null.'
      });

  } else {

    db.one(format('INSERT into comments(author_id, post_id, content, likes_nb, created, updated)' +
      'values(%1$L, %2$L, %3$L, 0, %4$L, %4$L) ' +
      'RETURNING comments.id AS comment_id',
      body.author_id, body.post_id, body.content, body.created))
      .then((response) => {

        res.status(201)
          .json({

            status: 'success',
            comment_id: response.comment_id,
            message: 'Inserted one comment from user ' + body.author_id + ' to post ' + body.post_id

          });

      })
      .catch(function (err) {

        return next(err);

      });

  }

}

function deleteComment(req, res, next) {

  const comment_id = parseInt(req.params.id);

  findCommentById(comment_id, (err, comment) => {

    if (err) {

      next(err);

    } else if (!req.user || parseInt(comment.author_id) !== req.user.id) {

      res.status(403)
        .json({
          status: 'error',
          message: 'You are not authorized to delete this comment.'
        });

    } else {

      db.result(format('DELETE FROM comments WHERE id = %L', comment_id))
        .then(function (result) {
          /* jshint ignore:start */
          res.status(200)
            .json({
              status: 'success',
              message: `Removed comment ${comment_id}`
            });
          /* jshint ignore:end */
        })
        .catch(function (err) {
          return next(err);
        });

    }

  });

}

function findCommentById(id, cb) {

  db.oneOrNone(format('SELECT * FROM comments WHERE id = %L', id))

    .then((data) => {

      return cb(null, data);

    })
    .catch((err) => {

      return cb(err, null);

    })

}

function getComment(req, res, next) {

  const comment_id = parseInt(req.params.id);

  if (comment_id) {

    db.any(format('SELECT comments.id, comments.content, comments.author_id, comments.likes_nb, comments.post_id, comments.created, comments.updated, ' +
      'users.login, users.firstname, users.lastname ' +
      'FROM comments INNER JOIN users ON comments.author_id = users.id ' +
      'WHERE comments.id = %L',
      comment_id))

      .then((data) => {

        if (Array.isArray(data)) {

          return (serializeComments(data));

        } else {

          return (serializeComment(data));

        }

      })
      .then((data) => {

        res.status(200)
          .json({

            status: 'success',
            data,
            message: 'Retrieved comment'

          });

      })
      .catch(function (err) {

        return next(err);

      });

  } else {

    res.status(400)
      .json({

        status: 'error',
        message: 'id not provided.'

      });

  }
}

function getComments(req, res, next) {

  if (req.query.post_id) {

    const post_id = parseInt(req.query.post_id);

    db.any(format('SELECT comments.id, comments.content, comments.author_id, comments.likes_nb, comments.post_id, comments.created, comments.updated, ' +
      'users.login, users.firstname, users.lastname ' +
      'FROM comments INNER JOIN users ON comments.author_id = users.id ' +
      'WHERE comments.post_id = %L ' +
      'ORDER BY comments.created ASC',
      post_id))

      .then((data) => {

        if (Array.isArray(data)) {

          return (serializeComments(data));

        } else {

          return (serializeComment(data));

        }

      })
      .then((data) => {

        res.status(200)
          .json({

            status: 'success',
            data,
            message: 'Retrieved comments'

          });

      })
      .catch(function (err) {

        return next(err);

      });

  } else {

    res.status(400)
      .json({

        status: 'error',
        message: 'post_id not provided.'

      });

  }
}

function serializeComments(data) {

  let newData = [];

  for (let i = 0, len = data.length; i < len; i++) {

    newData.push(serializeComment(data[i]));

  }

  return (newData);

}

function serializeComment(data) {

  return ({
    id: data.id,
    content: data.content,
    post_id: data.post_id,
    likes_nb: data.likes_nb,
    created: data.created,
    updated: data.updated,
    user: {
      id: data.author_id,
      login: data.login,
      firstname: data.firstname,
      lastname: data.lastname
    }
  });

}

function updateComment(req, res, next) {

  const comment_id = parseInt(req.params.id);
  const now = new Date();

  let content = req.body.content ? req.body.content : null;

  findCommentById(comment_id, (err, comment) => {

    if (err) {

      next(err);

    } else if (!req.user || parseInt(comment.author_id) !== req.user.id) {

      res.status(403)
        .json({
          status: 'error',
          message: 'You are not authorized to delete this comment.'
        });

    } else if (!content) {

      res.status(400)
        .json({
          status: 'error',
          message: 'content can not be null.'
        });

    } else {

      db.none(format('UPDATE comments SET content=COALESCE(%1$L, content), updated=%2$L WHERE id=%3$L',
        content, now, comment_id))
        .then(() => {

          res.status(200)
            .json({

              status: 'success',
              message: 'Updated comment ' + comment_id

            });

        })
        .catch(function (err) {

          return next(err);

        });

    }

  });

}

// Add query functions
module.exports = {
  createComment,
  deleteComment,
  findCommentById,
  getComments,
  getComment,
  updateComment,
  serializeComment,
  serializeComments
};

File: go-api/queries/comments.go
package queries

import (
	"github.com/labstack/echo"
	"../types"
	"time"
	"net/http"
	"../db"
	"fmt"
	"strconv"
	"github.com/jackc/pgx"
	"github.com/jackc/pgx/pgtype"
)

func CreateComment(c echo.Context) error {

	now := time.Now()

	var body types.CommentForm; var err error
	c.Bind(&body)
	loggedId := getUserId(c)

	if body.Content == "" {
		return c.JSON(http.StatusBadRequest, echo.Map{
			"status":     "error",
			"message":    "content can not be null.",
		})
	} else if body.PostId == 0 {
		return c.JSON(http.StatusBadRequest, echo.Map{
			"status":     "error",
			"message":    "post_id can not be null.",
		})
	} else {

		err = db.App.Db.QueryRow("INSERT into comments(author_id, post_id, content, likes_nb, created, updated)"+
			"values($1, $2, $3, 0, $4, $4) "+
			"RETURNING comments.id AS comment_id",
			loggedId, body.PostId, body.Content, now).
			Scan(&body.Id)

		if err != nil || body.Id <= 0 {

			return c.JSON(http.StatusInternalServerError, echo.Map{
				"status":  "error",
				"message": err.Error(),
			})
		}

		return c.JSON(http.StatusCreated, echo.Map{
			"status":  "success",
			"comment_id": body.Id,
			"message": fmt.Sprintf("Inserted one comment for user %v to post %v.",
				loggedId, body.PostId),
		})
	}

}

func DeleteComment(c echo.Context) error {

	commentId, ok := strconv.ParseInt(c.Param("id"), 10, 64)
	loggedId := getUserId(c)

	if ok != nil {
		return c.JSON(http.StatusBadRequest, echo.Map{
			"status":   "error",
			"message":  "Id not provided",
		})
	}

	comment, err := findCommentById(commentId)

	if err != nil {
		if err != pgx.ErrNoRows {
			return c.JSON(http.StatusInternalServerError, echo.Map{
				"status":   "error",
				"message":  err.Error(),
			})
		} else {
			return c.JSON(http.StatusNotFound, echo.Map{
				"status":   "error",
				"message":  "Comment not found",
			})
		}
	} else if comment.AuthorId.Status != pgtype.Present || int64(comment.AuthorId.Int) != loggedId {

		return c.JSON(http.StatusForbidden, echo.Map{
			"status":   "error",
			"message":  "You can't delete this comment",
		})

	} else {

		_, err := db.App.Db.Exec("DELETE FROM comments WHERE id = $1", comment.Id.Int)

		if err != nil {

			return c.JSON(http.StatusInternalServerError, echo.Map{
				"status":  "error",
				"message": err.Error(),
			})

		} else {
			message := fmt.Sprintf("Removed comment %v", comment.Id.Int)
			return c.JSON(http.StatusOK, echo.Map{
				"status":  "success",
				"message": message,
			})
		}
	}
}

func findCommentById(id int64) (types.Comment, error) {

	var comment types.Comment

	err := db.App.Db.QueryRow("SELECT * FROM comments WHERE id = $1", id).
		Scan(&comment.Id, &comment.AuthorId, &comment.PostId, &comment.Content, &comment.Created,
			&comment.Updated, &comment.LikesNb)

	if err != nil {
		if err == pgx.ErrNoRows {
			return types.Comment{}, err
		} else {
			CheckErr(err)
			return types.Comment{}, err
		}
	}

	return comment, nil
}

func GetComment(c echo.Context) error {

	var err error
	var comment types.SingleCommentData
	commentId, ok := strconv.ParseInt(c.Param("id"), 10, 64)

	if ok != nil {
		return c.JSON(http.StatusBadRequest, echo.Map{
			"status":     "error",
			"message":    "id not specified.",
		})
	}

	err = db.App.Db.QueryRow("SELECT comments.id, comments.content, comments.author_id, comments.likes_nb, " +
		"comments.post_id, comments.created, comments.updated, users.login, users.firstname, users.lastname " +
		"FROM comments INNER JOIN users ON comments.author_id = users.id WHERE comments.id = $1", commentId).
		Scan(&comment.Id, &comment.Content, &comment.AuthorId, &comment.LikesNb, &comment.PostId, &comment.Created,
			&comment.Updated, &comment.User.Login, &comment.User.Firstname, &comment.User.Lastname)

	if err != nil && err != pgx.ErrNoRows {
		return c.JSON(http.StatusInternalServerError, echo.Map{
			"status":  "error",
			"message": err.Error(),
		})
	} else {
		message := fmt.Sprintf("Retrieved comment %v.", comment.Id)
		return c.JSON(http.StatusOK, echo.Map{
			"status":   "success",
			"data":     comment,
			"message":  message,
		})
	}
}

func GetComments(c echo.Context) error {

	var postId int64; var ok, err error; var results *pgx.Rows; var comments []types.SingleCommentDataResponse
	postId, ok = strconv.ParseInt(c.QueryParam("post_id"), 10, 64)

	if ok != nil  {
		return c.JSON(http.StatusBadRequest, echo.Map{
			"status":     "error",
			"message":    "post_id not provided.",
		})
	} else {
		results, err = db.App.Db.Query("SELECT comments.id, comments.content, comments.author_id, " +
			"comments.likes_nb, comments.post_id, comments.created, comments.updated, users.login, users.firstname, " +
			"users.lastname FROM comments INNER JOIN users ON comments.author_id = users.id " +
			"WHERE comments.post_id = $1 ORDER BY comments.created ASC", postId)
	}

	if err != nil && err != pgx.ErrNoRows {
		return c.JSON(http.StatusInternalServerError, echo.Map{
			"status":   "error",
			"message":  err.Error(),
		})
	}

	for results.Next() {
		var comment types.SingleCommentData
		if err = results.
			Scan(&comment.Id, &comment.Content, &comment.AuthorId, &comment.LikesNb, &comment.PostId,
				&comment.Created, &comment.Updated, &comment.User.Login, &comment.User.Firstname,
				&comment.User.Lastname); err != nil {
			return c.JSON(http.StatusInternalServerError, echo.Map{
				"status":   "error",
				"message":  err.Error(),
			})
		}

		var singleComment types.SingleCommentDataResponse

		if comment.Id.Status == pgtype.Present {
			singleComment.Id = int64(comment.Id.Int)
		}
		if comment.Content.Status == pgtype.Present {
			singleComment.Content = comment.Content.String
		}
		if comment.AuthorId.Status == pgtype.Present {
			singleComment.AuthorId = int64(comment.AuthorId.Int)
			singleComment.User.Id = int64(comment.AuthorId.Int)
		}
		if comment.LikesNb.Status == pgtype.Present {
			singleComment.LikesNb = int64(comment.LikesNb.Int)
		}
		if comment.PostId.Status == pgtype.Present {
			singleComment.PostId = int64(comment.PostId.Int)
		}
		if comment.Created.Status == pgtype.Present {
			singleComment.Created = &(comment.Created.Time)
		}
		if comment.Updated.Status == pgtype.Present {
			singleComment.Updated = &(comment.Updated.Time)
		}
		if comment.User.Login.Status == pgtype.Present {
			singleComment.User.Login = comment.User.Login.String
		}
		if comment.User.Firstname.Status == pgtype.Present {
			singleComment.User.Firstname = comment.User.Firstname.String
		}
		if comment.User.Lastname.Status == pgtype.Present {
			singleComment.User.Lastname = comment.User.Lastname.String
		}

		comments = append(comments, singleComment)
	}
	if err = results.Err(); err != nil {
		return c.JSON(http.StatusInternalServerError, echo.Map{
			"status":   "error",
			"message":  err.Error(),
		})
	}

	return c.JSON(http.StatusOK, echo.Map{
		"status":   "success",
		"data":     comments,
		"message":  "Retrieved friendships",
	})
}

func UpdateComment(c echo.Context) error {

	loggedId := getUserId(c)
	commentId, ok := strconv.ParseInt(c.Param("id"), 10, 64)
	now := time.Now()
	var body types.CommentForm; var err error
	c.Bind(&body)

	if ok != nil {
		return c.JSON(http.StatusBadRequest, echo.Map{
			"status":     "error",
			"message":    "id not specified.",
		})
	}

	comment, err := findCommentById(commentId)

	if err != nil {
		if err != pgx.ErrNoRows {
			return c.JSON(http.StatusInternalServerError, echo.Map{
				"status":   "error",
				"message":  err.Error(),
			})
		} else {
			return c.JSON(http.StatusNotFound, echo.Map{
				"status":   "error",
				"message":  "Post not found",
			})
		}
	} else if loggedId != int64(comment.AuthorId.Int) {
		return c.JSON(http.StatusForbidden, echo.Map{
			"status":   "error",
			"message":  "You can't update this comment",
		})
	} else if body.Content == "" {

		return c.JSON(http.StatusBadRequest, echo.Map{
			"status":  "error",
			"message": "content can not be null.",
		})

	} else {

		_, err = db.App.Db.Exec("UPDATE comments SET content=COALESCE($1, content), updated=$2 WHERE id=$3",
		body.Content, now, commentId)

		if err != nil {
			return c.JSON(http.StatusInternalServerError, echo.Map{
				"status":  "error",
				"message": err.Error(),
			})
		} else {
			message := fmt.Sprintf("Updated comment %v.", comment.Id)
			return c.JSON(http.StatusOK, echo.Map{
				"status":  "success",
				"message": message,
			})
		}
	}
}
